### PR TITLE
Add RenderShadowTask

### PR DESCRIPTION
**Pull request by zzmp**
Introduces a new render::Task to fetch, cull, and render to a shadowmap:
- Added ViewFrustum::getCorners to help define light frustums
- Created a LightStage class to contain model::Lights with their shadowmaps
- Added Shadows to on-screen stats
- Added shadows to DebugDeferredBuffer using a color buffer

This also introduces two new jobs, PipelineSortShape and DepthSortShape, to facilitate only setting the pipeline twice.

This task is not yet running. It will be added to the engine in a future commit.

Other fixes:
- Fixed bitsetters for ShapeKey
- Fixed setDepthRangeTransform
- Guard against unspecified Render.deferredDebugMode
- Move GL context sync to Engine::run

Still needed:
- The quality of the shadows is very low (when enabled), and a better sampling scheme will be necessary.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: c704a8d8c56f8455960e7cc96a0ac99cf25c24d1 available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ec496f1907c0e103db10f28513518612421d18b4 available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2158e48a1995bb5913d42fa56e016ba5b90609b available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by samcake** [comment]
YOu still have a warning that fails the linux/mac builds:
/data/jenkins/Jenkins/workspace/hifi-pull-request-ubuntu/libraries/render-utils/src/LightStage.cpp:30:14: warning: variable ‘orientation’ set but not used [-Wunused-but-set-variable]
         auto orientation = glm::quat(glm::mat3(direction, up, side));

File: libraries/render-utils/src/LightStage.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        orientation = glm::quat(glm::mat3(IDENTITY_UP, -IDENTITY_FRONT, IDENTITY_RIGHT));
+    } else {
+        auto side = glm::normalize(glm::cross(direction, IDENTITY_UP));
+        auto up = glm::normalize(glm::cross(side, direction));
+        auto orientation = glm::quat(glm::mat3(direction, up, side));
+    }
+    _frustum->setOrientation(orientation);
+
+    // Position the keylight frustum
+    _frustum->setPosition(viewFrustum->getPosition() - 21.0f * direction);
```
Comment (samcake) [inline libraries/render-utils/src/LightStage.cpp]: Magic Number 21 needs to be defined as a const variable

File: libraries/render-utils/src/LightStage.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    _view = _frustum->getView();
+    const Transform viewInverse{ _view.getInverseMatrix() };
+
+    viewFrustum->calculate();
+    //const auto nearCorners = viewFrustum->getCorners(0);
+    const auto nearClip = viewFrustum->getNearClip();
+    const auto nearCorners = viewFrustum->getCorners(nearClip - 1);
+    const auto farCorners = viewFrustum->getCorners(nearClip + 20);
```
Comment (samcake) [inline libraries/render-utils/src/LightStage.cpp]: Magic numbers here
do you mean to use the parameters zBack and zFront here ?
i think the names zBack and zFront or not ideal, i d use near and far again (in this order)
if you want the true z  (probably < 0) then with zNear > zFar 
if you want the depth (probably > 0) then with nearDepth < nearFar

**Comment by hifi-gustavo** [comment]
Build console output for commit: d6e5e2de815079a2f0d25617c77aa3b430cdf11a available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5c1c1e3a3b5600b97178e06d69105ad66f367f6d available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by zzmp** [comment]
Looking at the timers, *everything* in the shadow task is taking a long time. I think it is because there are many more shapes - nothing is being culled as "out of view" and so the sorting jobs are taking a few ms each.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7f3e102b5ed820206df47f99129cf0adcd4d4489 available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

File: libraries/render-utils/src/directional_light.slf [hunk 3]
```diff
@@ -12,9 +12,8 @@
 //  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
```
Comment (samcake) [inline libraries/render-utils/src/directional_light.slf]: WE need to introduce a new version of the lighting shader that uses SHadows ( vs not, the current ones)

File: libraries/render-utils/src/directional_ambient_light.slf [hunk 4]
```diff
@@ -12,9 +12,8 @@
 //  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
```
Comment (samcake) [inline libraries/render-utils/src/directional_ambient_light.slf]: WE need to introduce a new version of the lighting shader that uses SHadows ( vs not, the current ones)

File: libraries/render-utils/src/directional_skybox_light.slf [hunk 5]
```diff
@@ -12,9 +12,8 @@
 //  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
```
Comment (samcake) [inline libraries/render-utils/src/directional_skybox_light.slf]: WE need to introduce a new version of the lighting shader that uses SHadows ( vs not, the current ones)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6434d983657cf47b7cf09dd7e32dfa9b02f6f15b available at: https://gist.github.com/b805707802a6b5b50b77<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6837.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6837.dmg

**Comment by samcake** [comment]
Checked, no impact on the rendering framerate

**Pull request merged**